File: JAVASCRIPT/classes.js
// // Create a Person constructor function:
// // It should take name and age as arguments.
// // It should initialize the object's name and age properties with the given arguments.
// // Add a greet method to the Person prototype that logs a greeting message including the person's name.
// class Person{
//     constructor(name, age){
//         this.name = name;
//         this.age = age;
//     }

//     greet(){
//       return "Greetings," + this.name
//     }
// };

// const firstPerson = new Person("Asha", 23);
// console.log(firstPerson.greet());

// // Create a Rectangle class:
// // It should have a constructor that takes width and height as arguments.
// // It should initialize the object's width and height properties.
// // Add methods to calculate the area and perimeter of the rectangle.
//  class Rectangle{
//     constructor(width, height){
//         this.width = width;
//         this.height = height;
//     }

//     perimeter(){
//         return 2 * this.width + 2 * this.height 
//     }

//     area(){
//         return this.width * this.height
//     }
//  }

//  const rectangle = new Rectangle(23,17)
//  console.log(rectangle.perimeter());
//  console.log(rectangle.area());


// // Create a BankAccount class:
// // It should have a constructor that takes accountNumber and balance as arguments.
// // It should initialize the object's accountNumber and balance properties.
// // Add methods to deposit and withdraw money, ensuring the balance doesn't go below zero.
// class Account{
//     constructor(accNumber, balance){
//         this.accNumber = accNumber;
//         this.balance = balance;
//     }

//     deposit(amount){
//         const totalAmount = this.balance + amount
//         if (amount < 0){
//             return amount + "cannot ne deposited"
//         } else {
//             return "Confirmed. You have successfully deposited" + " " + amount + " " + "Your new balance is" + " " + totalAmount
//         }

//     }

//     withdrawal(amount){
//         const remainingBalance = this.balance - amount
//         if(amount < 0){
//             "Cannot withdraw " + amount
//         } else if(amount > 0 && amount < this.balance){
//             return "Confirmed. You have successfully withdrawn" + " " + amount +  " " +"You new balance is" +" " + remainingBalance
//         }
//     }
// }

// const myAcc = new Account("Leley Kaku", 45362)
// console.log(myAcc.deposit(4321));
// console.log(myAcc.withdrawal(23160));


// // Create a Student class that inherits from the Person class:
// // It should have a constructor that takes name, age, and studentId as arguments.
// // It should call the Person constructor to initialize the name and age properties.
// // It should initialize the object's studentId property.
// // Add a method to display the student's information, including their name, age, and student ID.

// class Student{
//     constructor (name, age, studentId){
//         this.name = name;
//         this.age = age;
//         this.studentId = studentId;
//     }

  
// }

// // Create a Car class:
// // It should have properties like make, model, year, and speed.
// // Add methods to accelerate and brake, which modify the speed property.
// // Add a method to display the car's current speed.

// class Car{
//     constructor(make, model, year, speed){
//         this.make = make;
//         this.model = model;
//         this.year = year;
//         this.speed = speed;
//     };

//     accelerate(){
//        this.speed += 100;
//        return this.speed;
//     };

//     brake(){
//         this.speed -= this.speed;
//         return this.speed;
//     };
// };

// const newCar = new Car("Mazda", "CX-5", 2023, 187);
// console.log(newCar.accelerate());
// console.log(newCar.brake());

// // Factory function:
// // Create a factory function that creates objects representing different shapes (e.g., circle, 
// // square, triangle) based on the input parameters. Each shape object should have a method to
// //  calculate its area.
// class Factory{
//     constructor(shape, height, width, length, radius){
//         this.shape = shape;
//         this.height = height;
//         this.width = width;
//         this.length = length;
//         this.radius = radius;
    
//     };

//     rectangleArea(){
//         return this.length * this.width;
//     };

//     rectanglePerimeter(){
//         return this.length + this.width;
//     };

//     triangleArea(){
//         return this.length * this.height;
//     };

//     trianglePerimeter(){
//         return this.length + this.width + this.height;
//     };

//     circleArea(){
//         const pi = 3.142
//         return pi * this.radius ** 2
//     };

//     circlePerimeter(){
//         const pi = 3.142
//         return 2 * pi * this.radius * 2
//     };

// };

// const retangle = new Factory("rectangle",23, 12, 21, 28);
// console.log(retangle.rectangleArea());
// console.log(retangle.rectanglePerimeter());

// const triangular = new Factory("triangle", 23, 12, 21, 28);
// console.log(triangular.triangleArea());
// console.log(triangular.trianglePerimeter());

// const circular = new Factory("circle", 23, 12, 21, 28);
// console.log(circular. circleArea());
// console.log(circular.circlePerimeter());


// // 12. University Class with Department Management
// // Write a JavaScript program that creates a class called University with properties for 
// // university name and departments. Include methods to add a department, remove a department,
// // and display all departments. Create an instance of the University class and add and remove
// // departments.
// class University{
//     constructor(name){
//         this.name = name;
//         this.departments = [];
//     };

//     addDepartment(departmentList){
//         return this.departments.push(departmentList);

//     };

//     removeDepartment(){
//         return this.departments.shift();
//     };

//     displayDepartments(){
//         this.departments.forEach(department => {
//             console.log(`The ${this.name} has the departments of ${department}.`);
//         });
//     };

// };

// const uni1 = new University("University Of Nairobi");
// console.log(uni1.addDepartment(["Law", "Education", "Sciences", "Medicine", "Economy", "Design", "Model" ]));
// console.log(uni1.removeDepartment());
// console.log(uni1.displayDepartments());


// // Create a CustomerOrder class with properties: orderId (string), items (array of objects with 
// // name, quantity, price), and status (string). Add a method calculateTotal() that returns the
// //  total order amount. Write an async method processPayment() that simulates payment with a 
// // Promise that resolves after 2 seconds. After calling the method, change the status to "paid" 
// // and print a success message.

// class CustomerOrder{
//     constructor(orderId, status){
//         this.orderId = orderId;
//         this.status = status;
//         this.items = [{}];
//     };

//     calculateTotal() {
//         let total = 0;
//         items.map( item => {
//           total += item.quantity * item.price;  
//         });
//         return `The total cost is ${total}`;
//     };

//     async processPayment() {
//     console.log(`Processing payment for order ${this.orderId}`);
//     await new Promise(resolve => setTimeout(resolve, 2000));
//     this.status = "unpaid";
//     if(this.status === "paid"){
//         console.log(`Payment successful for order ${this.orderId}. Order status: ${this.status}`);
//     } else {
//         console.log(`Payment not successful for order ${this.orderId}. Order status: ${this.status}`);
//     };
    
//   };
// };


// const order = new CustomerOrder("1234HTD5")
// let items = [
//   { name: "watermelon", quantity: 2, price: 250 },
//   { name: "Banana", quantity: 3, price: 140 },
//   { name: "Orange", quantity: 5, price: 70 }
// ];

// console.log(`Order Total: ${order.calculateTotal()}`);

// order.processPayment();



// // Create a TeamMember class that takes name, role, and an array of tasks (each task is an object 
// // with title and completed boolean). Write a prototype method completeTask(taskTitle) that marks 
// // a task as completed. Write another method checkProgress() that returns a Promise resolving to "
// // All tasks completed!" or rejecting with "Pending tasks remaining" based on the state of the tas
// // ks array.
// class TeamMember{
//     constructor(name, role, tasks){
//         this.name= name;
//         this.role= role;
//         this.tasks = tasks
//     };

//     async checkProgress() {

        
//     };

// };

// TeamMember.prototype.completeTask = function(taskTitle){
//       tasks.forEach(task => {
//         if(task === taskTitle){
//            tasks.completed = true;
//         };
//       });

// };



// // Build a Candidate class with properties: name, position, and interviews (array of objects with date, status). 
// // Add a method scheduleInterview(date) that pushes a new interview with status "pending". Then write an async 
// // function sendConfirmation() that returns a Promise that resolves after 1 second with a message "Interview 
// // confirmed with [name]", and log the message.
// class Candidate{
//     constructor(name, position,interviews){
//         this.name = name;
//         this.position = position;
//         this.interviews = interviews;
//     };
//     scheduleInterview(date){
//         this.interviews.push(date);
//         return(this.interviews);
        
          
//     };

//     async sendConfirmation() {
//         console.log(`Processing interview for ${this.name}`);
//     await new Promise((resolve) => setTimeout(resolve, 1000));
//     console.log(`Interview for ${this.name} has been confirmed for ${date.date}`);
        
//     };
// };
// let date = {date: "21/1/2025", status:"pending"}

// const firstCandidate = new Candidate("Liam","6",[{date: "21/1/2025", status:"pending"},{date: "21/1/2025", status:"pending"}]);
// firstCandidate.scheduleInterview(date)
// firstCandidate.sendConfirmation();


// Create a class named Rectangle.
// It should have a constructor that takes width and height and sets them as properties.
// Add a method calculateArea that returns the area of the rectangle.
// Add a method calculatePerimeter that returns the perimeter of the rectangle.

class Rectangles{
    constructor(width, height){
        this.width = width;
        this.height = height;
    };

    calculateArea(){
        const area = this.width * this.height
        return area;
        
    };

    calculatePerimeter = () =>{
        const perimeter = 2 * this.width + this.height
        return perimeter
    };


    isSquare = () => {
    if(this.width === this.height){
        return "true"
    } else {
        return "false"
    };
};
};

const rec1 = new Rectangles(12,12);
console.log(rec1.calculateArea());
console.log(rec1.calculatePerimeter());
console.log(rec1.isSquare());

// Create an instance of the Rectangle class.
// Set width to 5 and height to 10.
// Call the calculateArea, calculatePerimeter, and isSquare methods on the instance.
// Create another instance of the Rectangle class that is a square. Verify the isSquare method returns true.



const rectangleOne = new Rectangles(5, 10);
console.log(rectangleOne.calculateArea());
console.log(rectangleOne.calculatePerimeter());
console.log(rectangleOne.isSquare());

File: KOTLIN/classes.kt
fun main() {
     var person1 = Person("Alma",30,"Ethiopia")
    println(person1.printPersonDetails())

    var rectangle1 = Rectangle(7, 13)
    println(rectangle1.rectangleArea())

    var bmw = Car("BMW", "i9", 2024)
    println(bmw.carDetails())

    var firstStudent = Student("Robustus", 16, 'A')
    println(firstStudent.checkPromotion())

    var blossomsOfTheSsavannah = Book("Blossoms Of The Savannah", "CHIE Chani", 2020)
    println(blossomsOfTheSsavannah.bookDetails())

    var bankAccountOne = BankAccount("ABC1234cd", 40000, "Molly Green")
    println(bankAccountOne.deposit(3000))
    println(bankAccountOne.withdrawal(500))

    var acc2 = BankAccount("Bcv2614DR", 64239, "Luois Vuitton")
    println(acc2.withdrawal(1327))
    println(acc2.deposit(3217))

    var circle1 = Circle(12, 'h', 'k')
    println(circle1.circumfrennce())

    var newEmployee = Employee("Meana Zista", 32, "Receptionist")
    println(newEmployee.employeeData())

    var equilateralTriangle = Triangle(3,4,5)
    println(equilateralTriangle.trianglePerimeter())

    var bags = Product("Denri", 2311.5, 6)
    println(bags.totalCost())

    var dog = Animal("Dog", "bark")
    println(dog.animalSound())

    var anna = Customer("Anna", "anna@gmail.com","Adelaide, 216B")
    println(anna.message())

    var seven = MathUtils()
    println(seven.factorial(7))
}
//1. Write a Kotlin program that creates a class 'Person' with properties for name, age, and country.
// Include a function to print the person's details.
class Person(var name: String, var age: Int, var country: String){
    fun printPersonDetails(){
        println("Name: $name")
        println("Age:$ age")
        println("Country : $country")
    }
}

//2. Write a Kotlin program that creates a class 'Rectangle' with properties for width and height.
// Include a function to calculate the rectangle area.

class Rectangle(var width:Int, var length:Int){
    fun rectangleArea(){
        println(length * width)
    }
}

//3. Write a Kotlin program that creates a class 'Car' with properties for make, model, and year.
// Include a function to display car information.

class Car(var make: String, var model:String, var year: Int){
    fun carDetails(){
        println("$make")
        println("$model")
        println("$year")
    }
}

//4. Write a Kotlin program that creates a class 'Student' with properties for name, age, and grade.
// Include a function to check if the student is eligible for promotion.

class Student(var name:String, var age:Int, var grade:Char){
    fun checkPromotion(){
        if(grade == 'A'){
            println("Congratulations $name, you have been promoted to the next class")
        } else if(grade == 'B'){
            println("$name, Meeting expectations")
        }
        else{
            println("Invalid grade")
        }
    }
}

//5. Write a Kotlin program that creates a class 'Book' with properties for title, author, and publication
// year. Include a function to display book details.
class Book(var tittle:String, var author: String, var publicationYear:Int){
    fun bookDetails(){
        println("Tittle: $tittle")
        println("Author: $author")
        println("Year of publication: $publicationYear")
    }
}

//6. Write a Kotlin program that creates a class 'BankAccount' with properties for account number, balance,
// and account holder name. Include deposit and withdrawal functions.

class BankAccount(var accountNumber: String, var balance:Int, var accuntHolder:String){
    fun deposit(amount: Int){
        if (amount == 0){
            println("$amount cannot be deposited")
        } else if(amount > 0){
            var totalAmount = amount + balance
            println("Confirmed. You have successfully deposited $amount , your new balance is $totalAmount")
        }
    }

    fun withdrawal(money: Int){
        var newBalance = balance - money
        if(money == 0){
            println("Insufficient amount")
        } else if(money > 0 && money < balance){
            println("You have successfully withdrawn $money, Your new balance is $newBalance")
        }
    }
}

//7. Write a Kotlin program that creates a class 'Circle' with properties for radius and center coordinates.
// Include a function to calculate the circle circumference.

class Circle(var radius:Int, var coord1: Char, varcoord2:Char){
    fun circumfrennce(){
        var pi = 3.142
        var d = radius * 2
        var circleCircumfrence = 2 * pi * d
        println(circleCircumfrence)
    }
}

//8. Write a Kotlin program that creates a class 'Employee' with properties for name, age, and designation.
// Include a function to display employee details.

class Employee(var name:String, var age:Int, var designation:String){
    fun employeeData(){
        println("Employee Name: $name")
        println("Employee Age: $age")
        println("Employee Designation: $designation")

    }
}

//9. Write a Kotlin program that creates a class 'Triangle' with side length properties.
// Include a function to calculate the triangle perimeter.
class Triangle(var side1:Int, var side2:Int, var side3:Int){
    fun trianglePerimeter(){
        var perimeter = side1 + side2 + side3
        println(perimeter)
    }
}

//10. Write a Kotlin program that creates a class 'Product' with properties for name, price, and quantity.
// Calculate the total cost of the product with a function.
class Product(var name:String, var price: Double, var quantity:Int){
    fun totalCost(){
        var totalAmount = price * quantity
        println(totalAmount)
    }
}

//11. Write a Kotlin program that creates a class 'Animal' with properties for name and sound.
// Include a function to make the animal's sound.
class Animal(var name: String, var sound:String){
    fun animalSound(){
        println("$name: $sound")
    }
}

//12. Write a Kotlin program that creates a class 'Customer' with properties for name, email, and
// address. Include a function to send a welcome email to the customer.
class Customer(var name:String, var email:String, var address:String){
    fun message(){
        println("Welcome, $name")
    }
}

//13. Write a Kotlin program that creates a class 'MathUtils' with static functions to calculate the
// factorial, square root, and cube root of a number.

class MathUtils() {
    fun factorial(num: Int): Long {
        if (num === 0) {
            return 1
        } else {
            return num * factorial(num - 1)
        }

    }

}